File: tic_tac_cs/Game.cs
namespace tic_tac_cs
{
    interface IGame
    {
        void SwitchCurrentPlayer();
        void Display();
        bool IsWinner();
        bool IsDraw();
        bool IsOver();
        void MakeMove();
    }

    internal class Game : IGame
    {
        private Player player1;
        private Player player2;
        private Player? currentPlayer;
        private Player? winner;
        private Board board;

        public Game(Player player1, Player player2)
        {
            this.player1 = player1;
            this.player2 = player2;
            currentPlayer = player1;
            winner = null;
            board = new();
        }

        public Player? CurrentPlayer
        {
            get { return currentPlayer; }
        }

        public Player? Winner
        {
            get { return winner; }
            set { winner = value; }
        }

        public Board Board
        {
            get { return board; }
        }

        public void Display()
        {
            board.Display();
        }

        public bool IsDraw()
        {
            return board.IsFull();
        }

        public bool IsOver()
        {
            if (IsWinner())
            {
                winner = currentPlayer;
                board.SetRemainingCells();
                Utils.ClearTerminal();
                Display();
                Console.WriteLine($"Game over. {winner?.Name} is the winner!");
                return true;
            }

            if (IsDraw())
            {
                winner = null;
                currentPlayer = null;
                Console.WriteLine("Game over. It is a draw!");
                return true;
            }

            return false;
        }

        public bool IsWinner()
        {
            return board.CheckWinner();
        }

        public void MakeMove()
        {
            if (currentPlayer == null)
            {
                throw new Exception("Current player not found");
            }

            while (true)
            {
                try
                {
                    string[] input = Utils.GetUserInput(currentPlayer.Name);
                    var (row, col) = Utils.InputToCoords(input);

                    board.SetCell(row, col, currentPlayer.Mark);
                    break;
                } catch (Exception e)
                {
                    Console.WriteLine(e.Message);
                }
            }
        }

        public void SwitchCurrentPlayer()
        {
            if (currentPlayer != null)
            {
                if (currentPlayer.Equals(player1))
                {
                    currentPlayer = player2;
                }
                else if (currentPlayer.Equals(player2))
                {
                    currentPlayer = player1;
                }
                
            }
        }
    }
}

File: tic_tac_fs/Game.fs
module Game

open System

type Game(p1: Player.Player, p2: Player.Player) =
    let mutable _currentPlayer: Option<Player.Player> = Some(p1)
    let mutable _winner: Option<Player.Player> = None

    member val player1: Player.Player = p1
    member val player2: Player.Player = p2
    member val board: Board.Board = Board.Board()

    member this.currentPlayer with get() = _currentPlayer and set(value) = _currentPlayer <- value
    member this.winner with get() = _winner and set(value) = _winner <- value

    member this.SwitchCurrentPlayer() =
        match this.currentPlayer with
        | Some(p) when p = p2 -> this.currentPlayer <- Some(p1)
        | Some(p) when p = p1 -> this.currentPlayer <- Some(p2)
        | _ -> ()
    
    member this.Display() =
        this.board.Display()

    member this.IsWinner() : bool =
        this.board.CheckWinner()

    member this.IsDraw() : bool =
        this.board.IsFull()

    member this.IsOver() : bool =
        if this.IsWinner() then
            this.winner <- this.currentPlayer
            this.board.SetRemainingCells()
            Utils.ClearTerminal()
            this.Display()
            match this.winner with
            | Some(winner) -> Console.WriteLine($"Game over! {winner.name} is the winner!")
            | None -> raise (new Exception("Error retrieving winner"))

            true
        elif this.IsDraw() then
            this.winner <- None
            this.currentPlayer <- None
            Console.WriteLine("Game over. It is a draw!")
            true
        else
            false

    member this.MakeMove() =
        let mutable Break = true
        match this.currentPlayer with
        | None -> raise (new Exception("Current player not found"))
        | Some (player) ->
            while Break do
                try
                    let (row, col) = Utils.GetUserInput(player.name)
                    this.board.SetCell(row, col, player.mark)
                    Break <- false
                with
                | e -> Console.WriteLine(e.Message)